In [ ]:
!pip install googlemaps

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Preparing metadata (setup.py) ... done
  Created wheel for googlemaps: filename=googlemaps-4.10.0-py3-none-any.whl size=40717 sha256=d262e0eee8342416d88b892d2eb7382dfd287312de8d79b31fcb3781340124fd
  Stored in directory: /root/.cache/pip/wheels/d9/5f/46/54a2bdb4bcb07d3faba4463d2884865705914cc72a7b8bb5f0
Successfully built googlemaps


In [ ]:
import googlemaps

# Crea una instancia del cliente de Google Maps
client = googlemaps.Client(key='')

# Define la ciudad que deseas consultar
ciudad = "San Francisco"

# Realiza la consulta a la API de Google Maps
geocode_result = client.geocode(ciudad)

# Muestra los resultados de la consulta
print(geocode_result)


[{'address_components': [{'long_name': 'San Francisco', 'short_name': 'SF', 'types': ['locality', 'political']}, {'long_name': 'San Francisco County', 'short_name': 'San Francisco County', 'types': ['administrative_area_level_2', 'political']}, {'long_name': 'California', 'short_name': 'CA', 'types': ['administrative_area_level_1', 'political']}, {'long_name': 'United States', 'short_name': 'US', 'types': ['country', 'political']}], 'formatted_address': 'San Francisco, CA, USA', 'geometry': {'bounds': {'northeast': {'lat': 37.9298239, 'lng': -122.28178}, 'southwest': {'lat': 37.6398299, 'lng': -123.1328309}}, 'location': {'lat': 37.7749295, 'lng': -122.4194155}, 'location_type': 'APPROXIMATE', 'viewport': {'northeast': {'lat': 37.812, 'lng': -122.3482}, 'southwest': {'lat': 37.70339999999999, 'lng': -122.527}}}, 'place_id': 'ChIJIQBpAG2ahYAR_6128GcTUEo', 'types': ['locality', 'political']}]


In [ ]:
import googlemaps
import pandas as pd

# Creamos una instancia del cliente de Google Maps con nuestra clave de API
client = googlemaps.Client(key='AIzaSyAslvs99bRqZ5uCLfPRDcp61VnxSUzVCUA')

# Especificamos la ciudad que queremos consultar
ciudad = "San Francisco"

# Realizamos la consulta a la API de Google Maps
geocode_result = client.geocode(ciudad)

# Convertimos la respuesta en un objeto DataFrame de pandas
df = pd.json_normalize(geocode_result)

# Mostramos el DataFrame en la consola
print(df)

                                  address_components       formatted_address  \
0  [{'long_name': 'San Francisco', 'short_name': ...  San Francisco, CA, USA   

                      place_id                  types  \
0  ChIJIQBpAG2ahYAR_6128GcTUEo  [locality, political]   

   geometry.bounds.northeast.lat  geometry.bounds.northeast.lng  \
0                      37.929824                     -122.28178   

   geometry.bounds.southwest.lat  geometry.bounds.southwest.lng  \
0                       37.63983                    -123.132831   

   geometry.location.lat  geometry.location.lng geometry.location_type  \
0              37.774929            -122.419415            APPROXIMATE   

   geometry.viewport.northeast.lat  geometry.viewport.northeast.lng  \
0                           37.812                        -122.3482   

   geometry.viewport.southwest.lat  geometry.viewport.southwest.lng  
0                          37.7034                         -122.527  


In [ ]:
import requests
from bs4 import BeautifulSoup

url = "https://es.wikipedia.org/wiki/Anexo:Ciudades_m%C3%A1s_pobladas_de_M%C3%A9xico"

response = requests.get(url)

soup = BeautifulSoup(response.text, "html.parser")

table = soup.find_all("table")[0]

rows = table.find_all("tr")[1:]

ciudades = []

for row in rows:
    cols = row.find_all("td")
    ciudad = cols[1].text.strip()
    ciudades.append(ciudad)

print(ciudades[:30])

['CDMX-01', 'BCN-01', 'MEX-01', 'GUA-01', 'PUE-01', 'CHH-01', 'JAL-01', 'JAL-02', 'NLE-01', 'MEX-02', 'CHH-02', 'YUC-01', 'ROO-01', 'COA-01', 'AGU-01', 'SON-01', 'BCN-02', 'SLP-01', 'SIN-01', 'QUE-01', 'MEX-03', 'MIC-01', 'MEX-04', 'TAM-01', 'COA-02', 'MEX-05', 'GRO-01', 'JAL-03', 'NLE-02', 'DUR-01']


In [ ]:
import csv
import requests
from bs4 import BeautifulSoup

# Definimos la lista de los países de interés
paises = ['Perú', 'Colombia', 'México', 'Uruguay', 'Argentina', 'Chile', 'Brasil', 'Ecuador', 'Estados Unidos']

# Creamos un archivo CSV para escribir los resultados
with open('ciudades.csv', mode='w', newline='') as file:
    writer = csv.writer(file)
    writer.writerow(['País', 'Ciudad'])

    # Iteramos sobre cada país
    for pais in paises:
        print(f'Extrayendo ciudades de {pais}...')
        url = f'https://www.citypopulation.de/es/{pais}/ciudades/'
        response = requests.get(url)

        # Creamos una instancia de BeautifulSoup para analizar la respuesta HTML
        soup = BeautifulSoup(response.content, 'html.parser')

        # Buscamos todas las filas de la tabla de ciudades
        table_rows = soup.find_all('tr')

        # Recorremos cada fila para extraer la información de la ciudad
        for row in table_rows:
            # Obtenemos las celdas de la fila
            cells = row.find_all('td')

            # Si hay celdas en la fila, significa que contiene información de la ciudad
            if len(cells) > 0:
                # Extraemos el nombre de la ciudad
                ciudad = cells[1].text.strip()

                # Escribimos la ciudad en el archivo CSV junto con el país correspondiente
                writer.writerow([pais, ciudad])
                
print('Extracción completada.')

Extrayendo ciudades de Perú...
Extrayendo ciudades de Colombia...
Extrayendo ciudades de México...
Extrayendo ciudades de Uruguay...
Extrayendo ciudades de Argentina...
Extrayendo ciudades de Chile...
Extrayendo ciudades de Brasil...
Extrayendo ciudades de Ecuador...
Extrayendo ciudades de Estados Unidos...
Extracción completada.


In [ ]:
import googlemaps
import pandas as pd

# Creamos una instancia del cliente de Google Maps con nuestra clave de API
client = googlemaps.Client(key='AIzaSyAslvs99bRqZ5uCLfPRDcp61VnxSUzVCUA')

# Especificamos la ciudad que queremos consultar
ciudad = "San Francisco"

# Realizamos la consulta a la API de Google Maps
geocode_result = client.geocode(ciudad)

# Creamos un DataFrame de pandas para almacenar los valores de las columnas
df = pd.DataFrame(columns=['latitud', 'longitud', 'nombre_ciudad', 'estado', 'pais'])

# Extraemos los valores de las columnas que nos interesan de la respuesta de la API
latitud = geocode_result[0]['geometry']['location']['lat']
longitud = geocode_result[0]['geometry']['location']['lng']
nombre_ciudad = ''
estado = ''
pais = ''
for component in geocode_result[0]['address_components']:
    if 'locality' in component['types']:
        nombre_ciudad = component['long_name']
    elif 'administrative_area_level_1' in component['types']:
        estado = component['long_name']
    elif 'country' in component['types']:
        pais = component['long_name']

# Agregamos los valores al DataFrame
df = df.append({
    'latitud': latitud,
    'longitud': longitud,
    'nombre_ciudad': nombre_ciudad,
    'estado': estado,
    'pais': pais
}, ignore_index=True)

# Mostramos el DataFrame en la consola
print(df)

     latitud    longitud  nombre_ciudad      estado           pais
0  37.774929 -122.419415  San Francisco  California  United States


<ipython-input-4-fe12e095af61>:31: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append({


In [ ]:
import pandas as pd

# Leer archivo xlsx y cargarlo en un DataFrame
df = pd.read_excel('/content/ciudades .xlsx')

# Imprimir el contenido del DataFrame
print(df)

            ciudad
0             Lima
1         Arequipa
2         Trujillo
3         Chiclayo
4            Piura
..             ...
252       Portland
253  Oklahoma City
254      Las Vegas
255     Louisville
256      Baltimore

[257 rows x 1 columns]


In [ ]:
import pandas as pd
import googlemaps

# Cargamos el archivo XLSX que contiene la lista de ciudades en la columna 'ciudad'
df_ciudades = pd.read_excel('/content/ciudades .xlsx', usecols=['ciudad'])

# Creamos una instancia del cliente de Google Maps con nuestra clave de API
client = googlemaps.Client(key='AIzaSyAslvs99bRqZ5uCLfPRDcp61VnxSUzVCUA')

# Creamos un DataFrame de pandas para almacenar los valores de las columnas
df_resultados = pd.DataFrame(columns=['ciudad', 'latitud', 'longitud', 'nombre_ciudad', 'estado', 'pais'])

# Iteramos sobre las filas del DataFrame de ciudades
for index, row in df_ciudades.iterrows():
    ciudad = row['ciudad']
    
# Realizamos la consulta a la API de Google Maps para obtener la información de la ciudad
geocode_result = client.geocode(ciudad)

# Verificamos que se haya obtenido al menos un resultado
if geocode_result:
    # Extraemos los valores de las columnas que nos interesan de la respuesta de la API
    latitud = geocode_result[0]['geometry']['location']['lat']
    longitud = geocode_result[0]['geometry']['location']['lng']
    nombre_ciudad = ''
    estado = ''
    pais = ''
    for component in geocode_result[0]['address_components']:
        if 'locality' in component['types']:
            nombre_ciudad = component['long_name']
        elif 'administrative_area_level_1' in component['types']:
            estado = component['long_name']
        elif 'country' in component['types']:
            pais = component['long_name']

    # Agregamos los valores al DataFrame de resultados
    df_resultados = df_resultados.append({
        'ciudad': ciudad,
        'latitud': latitud,
        'longitud': longitud,
        'nombre_ciudad': nombre_ciudad,
        'estado': estado,
        'pais': pais
    }, ignore_index=True)
else:
    # Agregamos un registro con valores nulos para indicar que no se encontró información para esta ciudad
    df_resultados = df_resultados.append({
        'ciudad': ciudad,
        'latitud': None,
        'longitud': None,
        'nombre_ciudad': None,
        'estado': None,
        'pais': None
    }, ignore_index=True)
# Guardamos los resultados en un archivo CSV
df_resultados.to_csv('resultados.csv', index=False)

<ipython-input-19-cb10e545c528>:37: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df_resultados = df_resultados.append({


In [ ]:
import pandas as pd

# lee el archivo CSV y conviértelo en un DataFrame de Pandas
df = pd.read_csv('/content/Book1.csv')

# imprime las primeras filas del DataFrame
print(df.head())

  CIUDADES,latitud,longitud,nombre_ciudad,estado,pais 
0                                          Lima,,,,,  
1                                      Arequipa,,,,,  
2                                      Trujillo,,,,,  
3                                      Chiclayo,,,,,  
4                                         Piura,,,,,  


In [ ]:
import pandas as pd

# Lee el archivo .xlsx y conviértelo a DataFrame
df = pd.read_excel('/content/ciudades .xlsx')

# Escribe el DataFrame en un archivo .csv con separador ',' y sin índice
df.to_csv('sencilla.csv', sep=',', index=False)


In [ ]:
import re
import pandas as pd
import googlemaps

# Creamos una instancia del cliente de Google Maps con nuestra clave de API
client = googlemaps.Client(key='AIzaSyAslvs99bRqZ5uCLfPRDcp61VnxSUzVCUA')

# Creamos un DataFrame de pandas para almacenar los valores de las columnas
df_resultados = pd.DataFrame(columns=['nombre', 'latitud', 'longitud', 'nombre_ciudad', 'estado', 'pais'])

# Lista de enlaces con la estructura especificada
enlaces = ['https://www.google.com/maps/place/Bodega+Eddy/@-12.0949668,-77.0737302,17.25z/data=!4m12!1m6!3m5!1s0x9105c9a45bcabf01:0x8595acd154e9797b!2sBodega+Eddy!8m2!3d-12.0953987!4d-77.0717411!3m4!',
           'https://www.google.com/maps/place/Casa+Andina+Standard+Tacna/@-18.0122499,-70.2497691,17z/data=!3m1!4b1!4m5!3m4!1s0x910e1dbb194541c7:0x68b10cd0f3134f53!8m2!3d-18.0122499!4d-70.2475751']

# Iteramos sobre los enlaces
for enlace in enlaces:
    # Extraemos la coordenada del enlace utilizando expresiones regulares
    coordenadas = re.findall(r'@(-?\d+.\d+),(-?\d+.\d+)', enlace)[0]
    latitud = float(coordenadas[0])
    longitud = float(coordenadas[1])
    
    # Realizamos la consulta a la API de Google Maps para obtener la información de la ciudad
    reverse_geocode_result = client.reverse_geocode((latitud, longitud))[0]
    
    # Extraemos los valores de las columnas que nos interesan de la respuesta de la API
    nombre_ciudad = ''
    estado = ''
    pais = ''
    for component in reverse_geocode_result['address_components']:
        if 'locality' in component['types']:
            nombre_ciudad = component['long_name']
        elif 'administrative_area_level_1' in component['types']:
            estado = component['long_name']
        elif 'country' in component['types']:
            pais = component['long_name']
    
    # Agregamos los valores al DataFrame de resultados
    df_resultados = df_resultados.append({
        'nombre': reverse_geocode_result['formatted_address'],
        'latitud': latitud,
        'longitud': longitud,
        'nombre_ciudad': nombre_ciudad,
        'estado': estado,
        'pais': pais
    }, ignore_index=True)

# Imprimimos el DataFrame de resultados
print(df_resultados)

                                              nombre    latitud   longitud  \
0  Jirón Salaverry 381, Magdalena del Mar 15086, ... -12.094967 -77.073730   
1               Gral Deustua 175b, Tacna 23001, Peru -18.012250 -70.249769   

       nombre_ciudad             estado  pais  
0  Magdalena del Mar  Provincia de Lima  Peru  
1              Tacna              Tacna  Peru  


<ipython-input-22-9dac5d7ffa0c>:38: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df_resultados = df_resultados.append({
<ipython-input-22-9dac5d7ffa0c>:38: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df_resultados = df_resultados.append({


In [ ]:
import re
import googlemaps
import pandas as pd

# Lista de URLs con estructura dada
urls = ['https://www.google.com/maps/place/Bodega+Eddy/@-12.0949668,-77.0737302,17.25z/data=!4m12!1m6!3m5!1s0x9105c9a45bcabf01:0x8595acd154e9797b!2sBodega+Eddy!8m2!3d-12.0953987!4d-77.0717411!3m4!', 'https://www.google.com/maps/place/Plaza+de+Armas+de+Cusco/@-13.5159119,-71.9763226,17z/data=!3m1!4b1!4m5!3m4!1s0x9105c6ba7d247833:0xe51e1af3b38e06b!8m2!3d-13.5159616!4d-71.9740892']

# Lista de coordenadas extraídas de las URLs
coordenadas = []
for url in urls:
    # Extraemos las coordenadas con una expresión regular
    match = re.search(r'@(-?\d+\.\d+),(-?\d+\.\d+)', url)
    latitud = match.group(1)
    longitud = match.group(2)
    coordenadas.append((latitud, longitud))

# Creamos una instancia del cliente de Google Maps con nuestra clave de API
client = googlemaps.Client(key='AIzaSyAslvs99bRqZ5uCLfPRDcp61VnxSUzVCUA')

# Creamos un DataFrame de pandas para almacenar los valores de las columnas
df_resultados = pd.DataFrame(columns=['latitud', 'longitud', 'nombre_ciudad', 'estado', 'pais'])

# Iteramos sobre las coordenadas obtenidas de las URLs
for latitud, longitud in coordenadas:
    
    # Realizamos la consulta a la API de Google Maps para obtener la información de la ciudad
    geocode_result = client.reverse_geocode((latitud, longitud))
    
    # Extraemos los valores de las columnas que nos interesan de la respuesta de la API
    nombre_ciudad = ''
    estado = ''
    pais = ''
    for component in geocode_result[0]['address_components']:
        if 'locality' in component['types']:
            nombre_ciudad = component['long_name']
        elif 'administrative_area_level_1' in component['types']:
            estado = component['long_name']
        elif 'country' in component['types']:
            pais = component['long_name']

    # Agregamos los valores al DataFrame de resultados
    df_resultados = df_resultados.append({
        'latitud': latitud,
        'longitud': longitud,
        'nombre_ciudad': nombre_ciudad,
        'estado': estado,
        'pais': pais
    }, ignore_index=True)

# Imprimimos el DataFrame de resultados
print(df_resultados)

       latitud     longitud      nombre_ciudad             estado  pais
0  -12.0949668  -77.0737302  Magdalena del Mar  Provincia de Lima  Peru
1  -13.5159119  -71.9763226              Cusco              Cuzco  Peru


<ipython-input-23-7f3f3600b128>:42: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df_resultados = df_resultados.append({
<ipython-input-23-7f3f3600b128>:42: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df_resultados = df_resultados.append({


In [ ]:
import re
import pandas as pd
import googlemaps

# Creamos una instancia del cliente de Google Maps con nuestra clave de API
client = googlemaps.Client(key='AIzaSyAslvs99bRqZ5uCLfPRDcp61VnxSUzVCUA')

# Creamos un DataFrame de pandas para almacenar los valores de las columnas
df_resultados = pd.DataFrame(columns=['nombre', 'latitud', 'longitud', 'direccion_exacta', 'nombre_ciudad', 'estado', 'pais'])

# Lista de enlaces con la estructura especificada
enlaces = ['https://www.google.com/maps/place/Bodega+Eddy/@-12.0949668,-77.0737302,17.25z/data=!4m12!1m6!3m5!1s0x9105c9a45bcabf01:0x8595acd154e9797b!2sBodega+Eddy!8m2!3d-12.0953987!4d-77.0717411!3m4!',
           'https://www.google.com/maps/place/Casa+Andina+Standard+Tacna/@-18.0122499,-70.2497691,17z/data=!3m1!4b1!4m5!3m4!1s0x910e1dbb194541c7:0x68b10cd0f3134f53!8m2!3d-18.0122499!4d-70.2475751']

# Iteramos sobre los enlaces
for enlace in enlaces:
    # Extraemos la coordenada del enlace utilizando expresiones regulares
    coordenadas = re.findall(r'@(-?\d+.\d+),(-?\d+.\d+)', enlace)[0]
    latitud = float(coordenadas[0])
    longitud = float(coordenadas[1])
    
    # Realizamos la consulta a la API de Google Maps para obtener la información de la ciudad
    reverse_geocode_result = client.reverse_geocode((latitud, longitud))[0]
    
    # Extraemos los valores de las columnas que nos interesan de la respuesta de la API
    nombre_ciudad = ''
    estado = ''
    pais = ''
    for component in reverse_geocode_result['address_components']:
        if 'locality' in component['types']:
            nombre_ciudad = component['long_name']
        elif 'administrative_area_level_1' in component['types']:
            estado = component['long_name']
        elif 'country' in component['types']:
            pais = component['long_name']
    
    # Realizamos una solicitud de reverse geocoding para obtener la dirección exacta
    reverse_geocode_exact = client.reverse_geocode((latitud, longitud), result_type='street_address')[0]
    direccion_exacta = reverse_geocode_exact['formatted_address']
    
    # Agregamos los valores al DataFrame de resultados
    df_resultados = df_resultados.append({
        'nombre': reverse_geocode_result['formatted_address'],
        'latitud': latitud,
        'longitud': longitud,
        'direccion_exacta': direccion_exacta,
        'nombre_ciudad': nombre_ciudad,
        'estado': estado,
        'pais': pais
    }, ignore_index=True)

# Imprimimos el DataFrame de resultados
print(df_resultados)

                                              nombre    latitud   longitud  \
0  Jirón Salaverry 381, Magdalena del Mar 15086, ... -12.094967 -77.073730   
1               Gral Deustua 175b, Tacna 23001, Peru -18.012250 -70.249769   

                                    direccion_exacta      nombre_ciudad  \
0  Jirón Salaverry 381, Magdalena del Mar 15086, ...  Magdalena del Mar   
1               Gral Deustua 175b, Tacna 23001, Peru              Tacna   

              estado  pais  
0  Provincia de Lima  Peru  
1              Tacna  Peru  


<ipython-input-24-a3c9689a02ca>:42: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df_resultados = df_resultados.append({
<ipython-input-24-a3c9689a02ca>:42: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df_resultados = df_resultados.append({


In [ ]:
import re
import pandas as pd
import googlemaps

# Creamos una instancia del cliente de Google Maps con nuestra clave de API
client = googlemaps.Client(key='AIzaSyAslvs99bRqZ5uCLfPRDcp61VnxSUzVCUA')

# Creamos un DataFrame de pandas para almacenar los valores de las columnas
df_resultados = pd.DataFrame(columns=['nombre', 'direccion_exacta', 'latitud', 'longitud', 'nombre_ciudad', 'estado', 'pais'])

# Lista de enlaces con la estructura especificada
#enlaces = ['https://www.google.com/maps/place/Bodega+Eddy/@-12.0949668,-77.0737302,17.25z/data=!4m12!1m6!3m5!1s0x9105c9a45bcabf01:0x8595acd154e9797b!2sBodega+Eddy!8m2!3d-12.0953987!4d-77.0717411!3m4!',
 #          'https://www.google.com/maps/place/Casa+Andina+Standard+Tacna/@-18.0122499,-70.2497691,17z/data=!3m1!4b1!4m5!3m4!1s0x910e1dbb194541c7:0x68b10cd0f3134f53!8m2!3d-18.0122499!4d-70.2475751']

import pandas as pd

# Lee el archivo de Excel y guarda los enlaces en un DataFrame de Pandas
df = pd.read_excel('/content/enlaces.xlsx')

# Convierte los enlaces en una lista
enlaces = df['enlace'].tolist()


# Iteramos sobre los enlaces
for enlace in enlaces:
    # Extraemos la coordenada del enlace utilizando expresiones regulares
    coordenadas = re.findall(r'@(-?\d+.\d+),(-?\d+.\d+)', enlace)[0]
    latitud = float(coordenadas[0])
    longitud = float(coordenadas[1])
    
    # Realizamos la consulta a la API de Google Maps para obtener la información de la ciudad
    reverse_geocode_result = client.reverse_geocode((latitud, longitud))[0]
    
    # Extraemos los valores de las columnas que nos interesan de la respuesta de la API
    nombre_ciudad = ''
    estado = ''
    pais = ''
    direccion_exacta = ''
    for component in reverse_geocode_result['address_components']:
        if 'locality' in component['types']:
            nombre_ciudad = component['long_name']
        elif 'administrative_area_level_1' in component['types']:
            estado = component['long_name']
        elif 'country' in component['types']:
            pais = component['long_name']
        elif 'route' in component['types']:
            direccion_exacta += component['long_name'] + ' '
        elif 'street_number' in component['types']:
            direccion_exacta += component['long_name']
    
    # Agregamos los valores al DataFrame de resultados
    df_resultados = df_resultados.append({
        'nombre': reverse_geocode_result['formatted_address'],
        'direccion_exacta': direccion_exacta,
        'latitud': latitud,
        'longitud': longitud,
        'nombre_ciudad': nombre_ciudad,
        'estado': estado,
        'pais': pais
    }, ignore_index=True)

# Imprimimos el DataFrame de resultados
print(df_resultados)

<ipython-input-28-c2ea283d3e87>:52: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df_resultados = df_resultados.append({
<ipython-input-28-c2ea283d3e87>:52: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df_resultados = df_resultados.append({
<ipython-input-28-c2ea283d3e87>:52: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df_resultados = df_resultados.append({
<ipython-input-28-c2ea283d3e87>:52: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df_resultados = df_resultados.append({
<ipython-input-28-c2ea283d3e87>:52: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat 

                                                nombre  \
0    Jirón Salaverry 381, Magdalena del Mar 15086, ...   
1           Jr. San Martín 310, San Miguel 15086, Peru   
2             Jirón Yungay 244, San Miguel 15086, Peru   
3    Jirón Diego Ferre 247, Magdalena del Mar 15086...   
4    Jr. Mariscal Ramon Castilla 855, Magdalena del...   
..                                                 ...   
127                 Mz b14 Lote 28, Callao 07041, Peru   
128         2WM6+G69, San Martín de Porres 15113, Peru   
129           Av. Cesar Vallejo 1090, Lima 15332, Peru   
130  2V5M+W87 Filadelfia IV Park, San Martín de Por...   
131  2V5M+W87 Filadelfia IV Park, San Martín de Por...   

                      direccion_exacta    latitud   longitud  \
0                  381Jirón Salaverry  -12.094967 -77.073730   
1                 310Jirón San Martín  -12.092144 -77.077341   
2                     244Jirón Yungay  -12.093900 -77.076890   
3                247Jirón Diego Ferre  -12.0975

<ipython-input-28-c2ea283d3e87>:52: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df_resultados = df_resultados.append({
<ipython-input-28-c2ea283d3e87>:52: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df_resultados = df_resultados.append({
<ipython-input-28-c2ea283d3e87>:52: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df_resultados = df_resultados.append({


In [ ]:
df_resultados.to_csv('resultados.csv', index=False)

In [ ]:
# Guardar el DataFrame en un archivo de Excel
df_resultados.to_excel('resultados_tienda.xlsx', index=False)

In [ ]:
pip install ortools

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.4/16.4 MB 57.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 302.4/302.4 KB 34.3 MB/s eta 0:00:00
  Attempting uninstall: protobuf
    Found existing installation: protobuf 3.19.6
    Uninstalling protobuf-3.19.6:
      Successfully uninstalled protobuf-3.19.6
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
tensorflow 2.11.0 requires protobuf<3.20,>=3.9.2, but you have protobuf 4.22.1 which is incompatible.
tensorflow-metadata 1.12.0 requires protobuf<4,>=3.13, but you have protobuf 4.22.1 which is incompatible.
tensorboard 2.11.2 requires protobuf<4,>=3.9.2, but you have protobuf 4.22.1 which is incompatible.


In [ ]:
import pandas as pd
from geopy.distance import distance

# Lee los datos del archivo Excel
data = pd.read_excel('/content/resultados_tienda.xlsx')

# Selecciona solo las columnas de latitud y longitud
lat_long = data[['latitud', 'longitud']]

# Itera sobre todas las combinaciones de coordenadas y calcula la distancia
for i in range(len(lat_long)):
    for j in range(i+1, len(lat_long)):
        coord1 = (lat_long.iloc[i]['latitud'], lat_long.iloc[i]['longitud'])
        coord2 = (lat_long.iloc[j]['latitud'], lat_long.iloc[j]['longitud'])
        dist = distance(coord1, coord2).km
        print(f"La distancia entre coordenadas {i} y {j} es {dist} km")

Se han truncado las últimas 5000 líneas del flujo de salida.
La distancia entre coordenadas 31 y 82 es 12.646311036529882 km
La distancia entre coordenadas 31 y 83 es 12.646311036529882 km
La distancia entre coordenadas 31 y 84 es 18.99266189717752 km
La distancia entre coordenadas 31 y 85 es 9.713975476397106 km
La distancia entre coordenadas 31 y 86 es 9.713975476397106 km
La distancia entre coordenadas 31 y 87 es 9.644299431770934 km
La distancia entre coordenadas 31 y 88 es 3.8223631742484327 km
La distancia entre coordenadas 31 y 89 es 10.835632235636458 km
La distancia entre coordenadas 31 y 90 es 10.835632235636458 km
La distancia entre coordenadas 31 y 91 es 7.659384940387834 km
La distancia entre coordenadas 31 y 92 es 7.748136979401205 km
La distancia entre coordenadas 31 y 93 es 8.98210853563882 km
La distancia entre coordenadas 31 y 94 es 8.98210853563882 km
La distancia entre coordenadas 31 y 95 es 1.0160381314338178 km
La distancia entre coordenadas 31 y 96 es 9.344637567

In [ ]:
import pandas as pd
from geopy.distance import distance

# Lee los datos del archivo Excel
data = pd.read_excel('/content/resultados_tienda.xlsx')

# Selecciona solo las columnas de latitud y longitud
lat_long = data[['latitud', 'longitud']]

# Crea una matriz vacía para almacenar las distancias
num_locations = len(lat_long)
dist_matrix = [[0] * num_locations for i in range(num_locations)]

# Itera sobre todas las combinaciones de coordenadas y calcula la distancia
for i in range(num_locations):
    for j in range(i+1, num_locations):
        coord1 = (lat_long.iloc[i]['latitud'], lat_long.iloc[i]['longitud'])
        coord2 = (lat_long.iloc[j]['latitud'], lat_long.iloc[j]['longitud'])
        dist = distance(coord1, coord2).km
        
        # Almacena la distancia en la matriz de distancias
        dist_matrix[i][j] = dist
        dist_matrix[j][i] = dist

# Imprime la matriz de distancias
for row in dist_matrix:
    print(row)

[0, 0.5020874115504875, 0.3636402438280519, 0.2890301302135297, 0.6619578490793552, 0.9235299363516676, 1.0648098035868174, 0.8073265081568853, 0.14573670995367352, 0.5680768303221782, 0.27356554236072783, 0.505877966571188, 0.8664459798398132, 0.07598237898576088, 0.7983481448798204, 3.8203346778797926, 2.3400207626103087, 2.2644028189943963, 3.5334125524811584, 3.600649126591759, 3.6379857739698642, 2.557585723022862, 3.479653315391991, 4.18440772091943, 3.9702059258841476, 2.3949443500568384, 3.636598464816736, 3.25043098936231, 1.9444180121322372, 1.9444180121322372, 2.337182893319287, 2.574223693309278, 2.5512884207085595, 3.128063772842702, 3.128063772842702, 3.030721944422051, 3.030721944422051, 2.392676875650193, 2.868449123494702, 2.7713718589519503, 3.5709265668323, 1.7459445708018864, 3.031757896423383, 2.551008236036135, 1.8749104398201961, 7.829821501731557, 1.162908170500742, 1.2195198847756987, 1.9983793529884752, 1.9983793529884752, 2.7027122157038566, 2.74446815031337,

In [ ]:
dist_matrix

[[0,
  0.5020874115504875,
  0.3636402438280519,
  0.2890301302135297,
  0.6619578490793552,
  0.9235299363516676,
  1.0648098035868174,
  0.8073265081568853,
  0.14573670995367352,
  0.5680768303221782,
  0.27356554236072783,
  0.505877966571188,
  0.8664459798398132,
  0.07598237898576088,
  0.7983481448798204,
  3.8203346778797926,
  2.3400207626103087,
  2.2644028189943963,
  3.5334125524811584,
  3.600649126591759,
  3.6379857739698642,
  2.557585723022862,
  3.479653315391991,
  4.18440772091943,
  3.9702059258841476,
  2.3949443500568384,
  3.636598464816736,
  3.25043098936231,
  1.9444180121322372,
  1.9444180121322372,
  2.337182893319287,
  2.574223693309278,
  2.5512884207085595,
  3.128063772842702,
  3.128063772842702,
  3.030721944422051,
  3.030721944422051,
  2.392676875650193,
  2.868449123494702,
  2.7713718589519503,
  3.5709265668323,
  1.7459445708018864,
  3.031757896423383,
  2.551008236036135,
  1.8749104398201961,
  7.829821501731557,
  1.162908170500742,
  1.

In [ ]:
import pandas as pd
from geopy.distance import distance
from ortools.constraint_solver import pywrapcp
from ortools.constraint_solver import routing_enums_pb2


# Lee los datos del archivo Excel
data = pd.read_excel('/content/resultados_tienda.xlsx')

# Selecciona solo las columnas de latitud y longitud
lat_long = data[['latitud', 'longitud']]

# Calcula la matriz de distancias a partir de las coordenadas
num_locations = len(lat_long)
dist_matrix = [[0.0] * num_locations for i in range(num_locations)]
for i in range(num_locations):
    for j in range(i+1, num_locations):
        coord1 = (lat_long.iloc[i]['latitud'], lat_long.iloc[i]['longitud'])
        coord2 = (lat_long.iloc[j]['latitud'], lat_long.iloc[j]['longitud'])
        dist = distance(coord1, coord2).km
        dist_matrix[i][j] = dist
        dist_matrix[j][i] = dist

# Parámetros del problema
num_vehicles = 25

# Creamos el solver de OR-Tools
solver = pywrapcp.Solver('car-routing')

# Definimos la función de costo
def distance_callback(from_index, to_index):
    return int(dist_matrix[from_index][to_index] * 1000)

# Creamos el administrador del índice
#manager = pywrapcp.RoutingIndexManager(num_locations, num_vehicles, [0], [0])
manager = pywrapcp.RoutingIndexManager(num_locations, num_vehicles, [0] * num_vehicles, [0])
# Creamos el modelo de rutas
routing = pywrapcp.RoutingModel(manager)

# Añadimos la función de costo
transit_callback_index = routing.RegisterTransitCallback(distance_callback)
routing.SetArcCostEvaluatorOfAllVehicles(transit_callback_index)

# Restricciones
search_parameters = pywrapcp.DefaultRoutingSearchParameters()
search_parameters.first_solution_strategy = (
    routing_enums_pb2.FirstSolutionStrategy.PARALLEL_CHEAPEST_INSERTION)

# Resolvemos el problema
solution = routing.SolveWithParameters(search_parameters) 

# Imprimimos la solución
if solution:
    for vehicle_id in range(num_vehicles):
        index = routing.Start(vehicle_id)
        route_distance = 0
        while not routing.IsEnd(index):
            print(manager.IndexToNode(index))
            previous_index = index
            index = solution.Value(routing.NextVar(index))
            route_distance += routing.GetArcCostForVehicle(previous_index, index, vehicle_id)
        print(manager.IndexToNode(index))
        print(f"Distancia del vehículo {vehicle_id}: {route_distance/1000}km\n")
else:
    print('No se encontró solución')

In [ ]:
pip install ortools

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
import pandas as pd
from geopy.distance import distance

# Lee los datos del archivo Excel
data = pd.read_excel('/content/resultados_tienda.xlsx')

# Selecciona solo las columnas de latitud y longitud
lat_long = data[['latitud', 'longitud']]

# Crea una matriz vacía para almacenar las distancias
num_locations = len(lat_long)
dist_matrix = [[0] * num_locations for i in range(num_locations)]

# Itera sobre todas las combinaciones de coordenadas y calcula la distancia
for i in range(num_locations):
    for j in range(i+1, num_locations):
        coord1 = (lat_long.iloc[i]['latitud'], lat_long.iloc[i]['longitud'])
        coord2 = (lat_long.iloc[j]['latitud'], lat_long.iloc[j]['longitud'])
        dist = distance(coord1, coord2).km
        
        # Almacena la distancia en la matriz de distancias
        dist_matrix[i][j] = dist
        dist_matrix[j][i] = dist

# Imprime la matriz de distancias
for row in dist_matrix:
    print(row)

[0, 0.5020874115504875, 0.3636402438280519, 0.2890301302135297, 0.6619578490793552, 0.9235299363516676, 1.0648098035868174, 0.8073265081568853, 0.14573670995367352, 0.5680768303221782, 0.27356554236072783, 0.505877966571188, 0.8664459798398132, 0.07598237898576088, 0.7983481448798204, 3.8203346778797926, 2.3400207626103087, 2.2644028189943963, 3.5334125524811584, 3.600649126591759, 3.6379857739698642, 2.557585723022862, 3.479653315391991, 4.18440772091943, 3.9702059258841476, 2.3949443500568384, 3.636598464816736, 3.25043098936231, 1.9444180121322372, 1.9444180121322372, 2.337182893319287, 2.574223693309278, 2.5512884207085595, 3.128063772842702, 3.128063772842702, 3.030721944422051, 3.030721944422051, 2.392676875650193, 2.868449123494702, 2.7713718589519503, 3.5709265668323, 1.7459445708018864, 3.031757896423383, 2.551008236036135, 1.8749104398201961, 7.829821501731557, 1.162908170500742, 1.2195198847756987, 1.9983793529884752, 1.9983793529884752, 2.7027122157038566, 2.74446815031337,

In [ ]:
import pandas as pd
import numpy as np
import math

# Cargamos el archivo Excel con las ubicaciones de las tiendas
ubicaciones = pd.read_excel("/content/resultados_tienda.xlsx")

# Convertimos las coordenadas de latitud y longitud en radianes
ubicaciones["lat_rad"] = np.radians(ubicaciones["lat_rad"])
ubicaciones["lon_rad"] = np.radians(ubicaciones["lon_rad"])

# Función para calcular la distancia entre dos puntos en la superficie de la Tierra, utilizando la fórmula de Haversine
def haversine(lat1, lon1, lat2, lon2):
    R = 6371  # Radio de la Tierra en kilómetros
    dlat = lat2 - lat1
    dlon = lon2 - lon1
    a = math.sin(dlat/2)**2 + math.cos(lat1) * math.cos(lat2) * math.sin(dlon/2)**2
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1-a))
    d = R * c
    return d

# Función para encontrar el punto más cercano no visitado
def encontrar_mas_cercano(actual, no_visitados):
    mas_cercano = None
    distancia_mas_cercana = None
    for i, fila in no_visitados.iterrows():
        d = haversine(actual["lat_rad"], actual["lon_rad"], fila["lat_rad"], fila["lon_rad"])
        if distancia_mas_cercana is None or d < distancia_mas_cercana:
            mas_cercano = fila
            distancia_mas_cercana = d
    return mas_cercano, distancia_mas_cercana

# Función para calcular la ruta óptima utilizando el algoritmo del vecino más cercano
def calcular_ruta_optima(ubicaciones):
    # Empezamos por el primer punto de la lista de ubicaciones
    actual = ubicaciones.iloc[0]
    ruta = [actual]
    no_visitados = ubicaciones.iloc[1:].copy()
    while len(no_visitados) > 0:
        siguiente, distancia = encontrar_mas_cercano(actual, no_visitados)
        if distancia is None:
            break
        ruta.append(siguiente)
        no_visitados.drop(siguiente.name, inplace=True)
        actual = siguiente
    ruta.append(ubicaciones.iloc[0]) # Volvemos al punto de inicio
    return ruta

# Ejecutamos la función para calcular la ruta óptima
ruta_optima = calcular_ruta_optima(ubicaciones)

# Imprimimos la ruta óptima en orden de visita
for i, fila in enumerate(ruta_optima):
    print(f"{i+1}. Tienda {fila['nombre']} ({fila['lat_rad']}, {fila['lon_rad']})")

1. Tienda Jirón Salaverry 381, Magdalena del Mar 15086, Peru (-0.2110969935794025, -1.3451903587837877)
2. Tienda Comandante Espinar 350, Magdalena del Mar 15086, Peru (-0.21110888974358408, -1.3451888543099726)
3. Tienda Castilla 260, Lima 15086, Peru (-0.2110886125083344, -1.345212116058243)
4. Tienda Jirón Echenique 215, Lima 15086, Peru (-0.21108814301476564, -1.3452332851567406)
5. Tienda Jirón Yungay 244, San Miguel 15086, Peru (-0.21107836917095446, -1.3452455007161752)
6. Tienda Jr. San Martín 310, San Miguel 15086, Peru (-0.2110477194439602, -1.3452533861137357)
7. Tienda Jr. Mariscal Ramon Castilla 855, Magdalena del Mar 15086, Peru (-0.21099269095797407, -1.3451849395364603)
8. Tienda Int 17, Jr.Leoncio Prado 814, Magdalena del Mar 15086, Peru (-0.21101308338495436, -1.3451583529359645)
9. Tienda Av Antonio José de Sucre 1022, Magdalena del Mar 15086, Peru (-0.21099416227053347, -1.345113979685062)
10. Tienda Jirón Junin 408, Magdalena del Mar 15086, Peru (-0.210981759960868

In [ ]:
import pandas as pd
from sklearn.metrics.pairwise import haversine_distances
from math import radians

data = pd.read_excel('/content/resultados_tienda.xlsx')
data['lat_rad'] = data['lat_rad'].apply(radians)
data['lon_rad'] = data['lon_rad'].apply(radians)

dist_matrix = haversine_distances(data[['lat_rad', 'lon_rad']])
dist_df = pd.DataFrame(dist_matrix, columns=data.index, index=data.index)
new_data = pd.concat([data, dist_df], axis=1)

new_data.to_excel('nombre_de_tu_archivo_con_distancias.xlsx', index=False)

In [ ]:
import pandas as pd
import numpy as np
from sklearn.metrics.pairwise import haversine_distances
from math import radians

data = pd.read_excel('/content/resultados_tienda.xlsx')

# convertir las columnas latitud y longitud a radianes
data['lat_rad'] = data['lat_rad'].apply(radians)
data['lon_rad'] = data['lon_rad'].apply(radians)

# calcular la matriz de distancias
dist_matrix = haversine_distances(data[['lat_rad', 'lon_rad']])

# crear un nuevo DataFrame a partir del array de distancias
dist_df = pd.DataFrame(dist_matrix, columns=data.index, index=data.index)

# combinar los dos DataFrames
new_data = pd.concat([data, dist_df], axis=1)

# guardar el resultado en un archivo de Excel
new_data.to_excel('nombre_de_tu_archivo_con_distanciasv2.xlsx', index=False)

In [ ]:
import pandas as pd

# Cargar los datos desde el archivo de Excel
data = pd.read_excel('/content/datos_de_entrada_modelo_ml_cervezas.xlsx')

In [ ]:
# Verificar si los datos se han cargado correctamente
data.head()

# Verificar si hay algún valor faltante en los datos
data.isnull().sum()

nombre               0
direccion_exacta    31
lat_rad              0
lon_rad              0
nombre_ciudad        0
                    ..
127                  0
128                  0
129                  0
130                  0
131                  0
Length: 142, dtype: int64

In [ ]:
# Eliminar columnas no necesarias
data = data.drop(['nombre', 'direccion_exacta', 'nombre_ciudad', 'estado', 'pais'], axis=1)

In [ ]:
# Codificar variables categóricas
data_encoded = pd.get_dummies(data)

In [ ]:
from sklearn.preprocessing import StandardScaler

# Escalar los datos
scaler = StandardScaler()
data_scaled = scaler

In [ ]:
# División de datos en conjunto de entrenamiento y prueba
from sklearn.model_selection import train_test_split

X = data.drop(['nombre', 'direccion_exacta', 'lat_rad', 'lon_rad', 'nombre_ciudad', 'estado', 'pais', 'Calificación'], axis=1)
y = data['Calificación']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


KeyError: ignored

In [ ]:
# Importamos las librerías necesarias
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

# Cargamos el dataset
data = pd.read_excel('/content/datos_de_entrada_modelo_ml_cervezas.xlsx')

# Definimos las variables independientes (X) y la variable dependiente (y)
X = data.drop(['nombre', 'direccion_exacta', 'nombre_ciudad', 'estado', 'pais', 'PRECIO_CERVEZA'], axis=1)
y = data['PRECIO_CERVEZA']

# Dividimos los datos en conjunto de entrenamiento y conjunto de prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

X_train.columns = X_train.columns.astype(str)

# Creamos un modelo de regresión lineal y lo entrenamos con los datos de entrenamiento
regressor = LinearRegression()
regressor.fit(X_train, y_train)

# Hacemos predicciones con los datos de prueba
# Convertir nombres de columna a tipo string
X_train.columns = X_train.columns.astype(str)
X_test.columns = X_test.columns.astype(str)

# Hacemos predicciones con los datos de prueba
y_pred = regressor.predict(X_test)

# Evaluamos el modelo usando el error cuadrático medio (MSE) y el coeficiente de determinación (R2)
mse = mean_squared_error(y_test, y_pred)
r2 = regressor.score(X_test, y_test)

# Imprimimos los resultados
print("Error cuadrático medio:", mse)
print("Coeficiente de determinación (R2):", r2)

Error cuadrático medio: 2.06706777368577
Coeficiente de determinación (R2): -2.3498408478946406


El error cuadrático medio (MSE) de 2.067 indica que el modelo está cometiendo un error promedio de 2.067 en las predicciones de precios de cerveza. Sin embargo, el coeficiente de determinación (R2) de -2.349 indica que el modelo no está ajustándose bien a los datos y, en realidad, está realizando predicciones peores que simplemente adivinar la media de los precios de cerveza. Es posible que deba revisar su conjunto de datos o considerar la posibilidad de utilizar otro tipo de modelo de aprendizaje automático.

Parece que el modelo de regresión lineal no es capaz de predecir bien los precios de cerveza en función de las variables proporcionadas en el conjunto de datos que estás utilizando. Es posible que se necesite utilizar otro tipo de modelo o ajustar los parámetros del modelo actual para obtener mejores resultados.

Además, también es posible que el conjunto de datos no contenga suficiente información o que las variables proporcionadas no sean suficientemente relevantes para predecir los precios de la cerveza con precisión. En ese caso, podría ser necesario recopilar más datos o utilizar diferentes variables para mejorar la precisión del modelo.